File: src/tame_mt/ngrams.py
from __future__ import annotations


def char_ngrams(text: str, orders: tuple[int, ...] = (3, 4, 5)) -> frozenset[str]:
    if not text:
        return frozenset()
    min_order = min(orders)
    if len(text) < min_order:
        return frozenset({text})

    grams: set[str] = set()
    for order in orders:
        if order <= len(text):
            grams.update(text[i : i + order] for i in range(len(text) - order + 1))
    return frozenset(grams)

File: src/ngrams.rs
pub(crate) fn char_ngram_slices<'a>(text: &'a str, orders: &[usize]) -> Vec<&'a [u8]> {
    if text.is_empty() {
        return Vec::new();
    }

    let mut offsets: Vec<usize> = text.char_indices().map(|(idx, _)| idx).collect();
    offsets.push(text.len());
    let char_count = offsets.len() - 1;
    let min_order = orders.iter().copied().min().unwrap_or(1);
    if char_count < min_order {
        return vec![text.as_bytes()];
    }

    let mut grams: Vec<&[u8]> = Vec::new();
    for order in orders {
        if *order <= char_count {
            for start in 0..=(char_count - *order) {
                let byte_start = offsets[start];
                let byte_end = offsets[start + *order];
                grams.push(&text.as_bytes()[byte_start..byte_end]);
            }
        }
    }

    grams.sort_unstable();
    grams.dedup();
    grams
}
